docs/Lowering.md

```diff
@@ -0,0 +1,230 @@
+          }
+          unsigned Counter = VecSize >> 1;
+          while (Counter != 0) {
+            for (unsigned Idx = 0; Idx < Counter; ++Idx) {
+              Extracts[Idx] = BM->addBinaryInst(Op, ResultSType, Extracts[Idx << 1],
+                                                Extracts[(Idx << 1) + 1], BB);
+            }
+            Counter >>= 1;
+          }
+          if ((VecSize & 1) != 0) {
+            Extracts[0] = BM->addBinaryInst(Op, ResultSType, Extracts[0],
+                                            Extracts[VecSize - 1], BB);
+          }
+          return Extracts[0];
+        }
+
+
+3. In SPIRVRegularizeLLVMPass in SPIRVRegularizeLLVM.cpp, calls to LLVM intrinsics are replaced with a call to an emulation function.
+   The emulation function is created by LLVM API calls and will be translated to SPIRV. The calls to the emulation
+   functions and the emulation functions themselves will be translated to SPIRV.  After reverse translation, the calls to the emulation
+   functions and the emulation functions themselves will appear in the LLVM IR.
+
+   For example, calls to llvm.bswap.i16:
+
+        %ret = call i16 @llvm.bswap.i16(i16 %0)
+
+   will be re-directed to an emulation function:
+
+        %ret = call i16 @spirv.llvm_bswap_i16(i16 %0)
+
+   The emulation function is constructed by the translator in SPIRVRegularizeLLVM.cpp (note that this code
+   handles all types):
+
+            case Intrinsic::bswap: {
+              BasicBlock *EntryBB = BasicBlock::Create(M->getContext(), "entry", F);
+              IRBuilder<> IRB(EntryBB);
+              auto *BSwap = IRB.CreateIntrinsic(Intrinsic::bswap, Intrinsic->getType(),
+                                                F->getArg(0));
+              IRB.CreateRet(BSwap);
+              IntrinsicLowering IL(M->getDataLayout());
+              IL.LowerIntrinsicCall(BSwap);
+              break;
+            }
+
+   This will produce a function like:
+
+        define i16 @spirv.llvm_bswap_i16(i16 %0) {
+        entry:
+          %bswap.2 = shl i16 %0, 8
+          %bswap.1 = lshr i16 %0, 8
+          %bswap.i16 = or i16 %bswap.2, %bswap.1
+          ret i16 %bswap.i16
+        }
+
+   After forward translation the emulation calls and functions will appear in SPIRV:
+
+        8 Name 24 "spirv.llvm_bswap_i16"
+        ...
+        5 FunctionCall 8 26 24 22
+        ...
+        5 Function 8 24 0 23
+        3 FunctionParameter 8 25
+        2 Label 42
+        5 ShiftLeftLogical 8 44 25 43
+        5 ShiftRightLogical 8 45 25 43
+        5 BitwiseOr 8 46 44 45
+        2 ReturnValue 46
+
+   In reverse translation, the lowering is undone.  Calls are reverted to the original llvm.bswap.i16 intrinsic
+
+        %ret = call i16 @llvm.bswap.i16(i16 %0)
+
+   The emulation functions are deleted.
+
+   This is the next best way of lowering a function.  The functionality of the intrinsic is created by a call to LLVM's CreateIntrinsic,
```

Comment:
It's great to spell out pros and cons of each approach. But TBH I'd prefer to avoid saying "the next best way" or "least preferable way" as it might make a reader to lean towards one way of lowering instead of another despite the fact, that another approach might be more suitable in their case.

Reply:
Removed comments rating alternatives.  Let reader decide.